removed broken checking for dynamic overrides of matrix free methods

naylor-b opened the pull request:
### Summary

Checking for dynamic overrides of matrix free methods was broken in cases where decorators were applied to those methods.  That checking was removed, so now only 'static' checking is done, i.e., if the class overrides the method we will automatically detect it as before, but if the method is added dynamically, it's now the responsibility of the component author to set the `matrix_free` attribute for that component to True manually because the framework will not attempt to detect it.

The same also applies to the `_has_solve_nl` flag in ImplicitComponent which suffered from the same problem as the `matrix_free` flag.

### Related Issues

- Resolves #2735

### Backwards incompatibilities

In certain cases where `compute_jacvec_product`, `apply_linear`, or `solve_nonlinear` methods were added dynamically to a component, it's possible that the framework used to detect them correctly (in the absence of decorators) but now will not.  The fix for this is to set the component's `matrix_free` attribute to True manually.

### New Dependencies

None

swryan reviewed (commented):
Would this be considered an API change, and does itrequire documentation?

naylor-b commented:
OK I updated the docstring for System (where matrix_free is documented) and also added a little to the docstring for `overrides_method`, which we use to detect the 'typical' case.
